### PR TITLE
osd: fine-grained statistics of logical object space usage

### PR DESCRIPTION
**Pull request by xiexingguo**
To test this change, we create an image of 5GB and do rbd bench write of 100MB:
`./bin/rbd create bar -s 5120 && ./bin/rbd bench --io-type write --io-size 32K --io-total 100M --io-pattern rand  rbd/bar`

Below is the test result.

Was（**3191MB**）:

```
GLOBAL:
    SIZE       AVAIL      RAW USED     %RAW USED
    30911M     27052M        3859M         12.49
POOLS:
    NAME                  ID     USED      %USED     MAX AVAIL     OBJECTS
    rbd                   0      3191M     26.36         8914M        1174
    cephfs_data_a         1          0         0         8914M           0
    cephfs_metadata_a     2       2246         0         8914M          21
```

Now(about **100MB**):

```
GLOBAL:
    SIZE       AVAIL      RAW USED     %RAW USED
    30911M     27050M        3861M         12.49
POOLS:
    NAME                  ID     USED        %USED     MAX AVAIL     OBJECTS
    rbd                   0      101216k      1.10         8913M        1178
    cephfs_data_a         1            0         0         8913M           0
    cephfs_metadata_a     2          892         0         8913M          21
```

E.g., this change can make "osd pool set-quota max_bytes" work nicely.

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas Any comments?

**Comment by liewegas** [comment]
Interesting!  This is surprisingly simple.

In order to rely on this we probably need to update (deep?) scrub to verify the accuracy of the extents field.  Presumably by reading everything outside the range and asserting it is zeros...

**Comment by liewegas** [comment]
@jdurgin what do you think?

**Comment by jdurgin** [comment]
I like the idea, I'm afraid it will end up a little more complex though. e.g. what about rbd clusters that use the set_alloc_hint to fully allocate space for an object with the first write?

This would also need to handle other write ops, e.g. truncate, zero, rollback, delete, etc.

write_update_size_and_usage() is only called for write, writefull, and ops that are implemented in terms of those.

**Comment by liewegas** [comment]
Yeah, I like the idea. If we can capture the other ops with similarly concise updates I'm all for this.

We'd need to make the compat decode path assume the full range is allocated in order to handle upgrades, and make sure we only trigger the sub-extent updates only if require_osd_release >= MIMIC.

And scrub.

**Comment by xiexingguo** [comment]
This passed several round of QA tests:
http://pulpito.ceph.com/xxg-2017-09-24_03:06:16-rados-wip-object-logic-size-distro-basic-smithi/
http://pulpito.ceph.com/xxg-2017-09-24_11:00:41-rados-wip-object-logic-size-distro-basic-smithi/
http://pulpito.ceph.com/xxg-2017-09-24_05:09:32-rados-wip-object-logic-size-distro-basic-smithi/

@liewegas I believe this is ready for review!

**Comment by xiexingguo** [comment]
@liewegas ping

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -7094,14 +7148,58 @@ void PrimaryLogPG::write_update_size_and_usage(object_stat_sum_t& delta_stats, o
...
+    // no change
+    return;
+  }
+  if (oi.has_extents()) {
+    int64_t old_bytes = oi.extents.size();
+    if (truncate_size > oi.size) {
+      // trunc up
+      interval_set<uint64_t> to_add;
+      to_add.insert(oi.size, truncate_size - oi.size);
+      oi.extents.union_of(to_add);
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: should a simple oi.extents.insert(oi.size, truncate_size - oi.size) be sufficient?  no need for to_add and union_of

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -6883,8 +6917,28 @@ int PrimaryLogPG::_rollback_to(OpContext *ctx, ceph_osd_op& op)
...
+      if (rollback_to->obs.oi.has_extents()) {
+        ctx->delta_stats.num_bytes += rollback_to->obs.oi.extents.size();
+        // transfer extents map too
+        assert(obs.oi.has_extents());
+        // can't do swap here, which has side-effect...
+        for (interval_set<uint64_t>::const_iterator p =
+             rollback_to->obs.oi.extents.begin();
+             p != rollback_to->obs.oi.extents.end(); ++p) {
+          obs.oi.extents.insert(p.get_start(), p.get_len());
+        }
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: wouldn't obs.oi.extents = rollback_to->obs.oid.extents be sufficient?

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -5091,6 +5100,7 @@ void object_info_t::dump(Formatter *f) const
   f->dump_unsigned("expected_write_size", expected_write_size);
   f->dump_unsigned("alloc_hint_flags", alloc_hint_flags);
   f->dump_object("manifest", manifest);
+  f->dump_stream("logical_extents") << extents;
```
Comment (liewegas) [inline src/osd/osd_types.cc]: may as well make this structures.  open_array_section("extents"), open_object_section("extent"), dump_unsigned("offset", ...), ...

File: src/osdc/Objecter.h [hunk 4]
```diff
@@ -796,6 +799,13 @@ struct ObjectOperation {
 	  *out_truncate_seq = copy_reply.truncate_seq;
 	if (out_truncate_size)
 	  *out_truncate_size = copy_reply.truncate_size;
+        if (out_extents && !copy_reply.extents.empty()) {
+          for (interval_set<uint64_t>::const_iterator p =
+               copy_reply.extents.begin();
+               p != copy_reply.extents.end(); ++p) {
+            out_extents->insert(p.get_start(), p.get_len());
+          }
```
Comment (liewegas) [inline src/osdc/Objecter.h]: again, here, why not use operator=?

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
@liewegas  All comments addressed.

**Comment by xiexingguo** [comment]
changeset:
fix a typo in comment

**Comment by xiexingguo** [comment]
There are suspicious scrub-errors, need more testing and investigations.

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2017-09-29_22:57:51-rados-wip-object-logic-size-distro-basic-smithi/
http://pulpito.ceph.com/xxg-2017-09-29_10:09:01-rados-wip-object-logic-size-distro-basic-smithi/
http://pulpito.ceph.com/xxg-2017-09-29_13:00:30-rados-wip-object-logic-size-distro-basic-smithi/

**Pull request merged**